### PR TITLE
tickets/DM-33509B: fix test suite for kubernetes_asyncio

### PR DESCRIPTION
**Pull request by athornton**
This works as well as vanilla kubespawner does for me, and test_spawn_internal_ssl() fails the same way, returning `''` rather than `'302'`.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -480,8 +514,8 @@ def _exec_python_in_pod(kube_client, kube_ns, pod_name, code, kwargs=None, _retr
     print("Running {} in {}".format(code, pod_name))
     # need to create ws client to get returncode,
     # see https://github.com/kubernetes-client/python/issues/812
-    client = stream(
-        kube_client.connect_get_namespaced_pod_exec,
+    client = sync_stream(
+        sync_corev1api.connect_get_namespaced_pod_exec,
```
Comment (rra) [inline tests/conftest.py:518]: This deserves a comment for why you're using kubernetes here and not kubernetes_asyncio.

File: tests/conftest.py [hunk 2]
```diff
@@ -526,7 +560,7 @@ def exec_python_pod(kube_client, kube_ns):
 
 
 @pytest.fixture(scope="session")
```
Comment (rra) [inline tests/conftest.py:562]: `@pytest_asyncio.fixture`

File: tests/conftest.py [hunk 3]
```diff
@@ -180,57 +200,68 @@ def kube_client(request, kube_ns):
...
 
 
-def wait_for_pod(kube_client, kube_ns, pod_name, timeout=90):
+async def wait_for_pod(kube_client, kube_ns, pod_name, timeout=90):
     """Wait for a pod to be ready"""
     conditions = {}
     for i in range(int(timeout)):
-        pod = kube_client.read_namespaced_pod(namespace=kube_ns, name=pod_name)
+        p = await kube_client.read_namespaced_pod(namespace=kube_ns, name=pod_name)
+        pod = p
```
[the remark this answers is not in the document]
Comment (athornton) [reply tests/conftest.py:256]: Left over from trying to debug some sync/async issues.  Fixed now.

File: tests/conftest.py [hunk 4]
```diff
@@ -180,57 +211,63 @@ def kube_client(request, kube_ns):
...
-        for i in range(3):
+    t = asyncio.create_task(watch_kubernetes(client, kube_ns))
+
+    yield client
+
+    # Clean up at close
+    await client.stop_signal.put(1)
+    # Wait until tasks have been cancelled
+    #
+    # I do not understand why join() really really blocks everything and the
```
Comment (athornton) [inline tests/conftest.py:237]: @rra Please explain this to me like I am five.  It seems like this is what join() should be for, but if I call it, absolutely *everything* stops.

File: tests/conftest.py [hunk 5]
```diff
@@ -162,16 +177,32 @@ def watch_kubernetes(kube_client, kube_ns):
...
+                watch_logs(
+                    kube_client,
+                    obj,
+                ),
             )
-            t.start()
 
+        # Were we asked to stop?
+        if not stop_signal.empty():
+            break
```
Comment (rra) [inline tests/conftest.py:191]: Your problem is here: you need to pull the signal off the queue and mark it as finished, ie:
```python
if not stop_signal.empty():
    await stop_signal.get()
    stop_signal.task_done()
    break
```

File: tests/conftest.py (hunk 4 above)
Comment (rra) [reply tests/conftest.py:237]: The problem is that on the other end you were only checking whether the queue was empty but never pulling the item off the queue and then acknowledging it with `task_done`, which is what `join` waits for.

**Comment by athornton** [comment]
So, for posterity...it wasn't that, exactly.  It's that the async for was never progressing because it didn't have a new event, but in that case, why did the polling approach work?

ANYWAY, changing it to throwing cancels back and forth seems to work just fine.

File: tests/conftest.py [hunk 6]
```diff
@@ -25,16 +27,29 @@
...
+from kubernetes.config import load_kube_config as sync_load_kube_config
+from kubernetes.client import CoreV1Api as sync_CoreV1Api
+
 here = os.path.abspath(os.path.dirname(__file__))
 jupyterhub_config_py = os.path.join(here, "jupyterhub_config.py")
 
+# We do these to set up the synchronous client, needed for executing
+# python inside pods.
+sync_load_kube_config()
+sync_corev1api = sync_CoreV1Api()
```
Comment (rra) [inline tests/conftest.py:46]: I think it should work to just move this into the `kube_client` fixture, which would get rid of some top-level stuff that feels a bit unclean.

File: tests/conftest.py [hunk 7]
```diff
@@ -7,12 +8,13 @@
 import sys
 import tarfile
 import time
-from distutils.version import LooseVersion as V
+from packaging.version import Version as V
 from functools import partial
-from threading import Thread
+from threading import Thread, Event
```
Comment (rra) [inline tests/conftest.py:13]: I suspect at least `Event` is no longer needed, and possibly also not `Thread`?

File: tests/conftest.py [hunk 8]
```diff
@@ -125,53 +137,72 @@ def watch_logs(kube_client, pod_info):
...
+        await stop_signal.get()
+        for t in watch_task:
+            if watch_task[t] and not watch_task[t].done():
+                try:
+                    watch_task[t].cancel()
+                except asyncio.CancelledError:
+                    # Swallow these; they are what we expect.
+                    pass
+        # And re-raise so kube_client can finish cleanup
+        raise exc
```
Comment (rra) [inline tests/conftest.py:201]: You can just use `raise` here and drop the `as exc` from the line earlier, I believe.

File: tests/conftest.py [hunk 9]
```diff
@@ -180,57 +211,61 @@ def kube_client(request, kube_ns):
     - Hooks up kubernetes events and logs to pytest capture
     - Cleans up kubernetes namespace on exit
     """
-    load_kube_config()
+    await load_kube_config()
     client = shared_client("CoreV1Api")
+    stop_signal=asyncio.Queue()
```
Comment (rra) [inline tests/conftest.py:216]: This can go away now.

**Review by rra — APPROVED** [review]
Just a few minor notes but otherwise looks good to me.

**Pull request merged**